keel-titus-plugin/src/main/kotlin/com/netflix/spinnaker/keel/api/titus/cluster/TitusClusterHandler.kt

```diff
@@ -126,16 +127,45 @@ class TitusClusterHandler(
             else -> diff.upsertServerGroupJob() + resource.spec.deployWith.toOrcaJobProperties()
           }
 
+          var tags: Set<String> = emptySet()
+          val version = when {
+            diff.isCapacityOnly() -> null
+            else -> {
+              // calculate the version for the digest
+              tags = getTagsForDigest(desired.container, desired.location.account).filterNot { it == "latest" }.toSet()
+              if (tags.size == 1) {
+                tags.first()
+              } else {
+                log.debug("Container digest ${desired.container} has multiple tags: $tags")
+                // unclear which "version" to print if there is more than one, so use digest
+                desired.container.digest
+              }
+            }
+          }
+
+          val description = when (version) {
+            null -> "Resizing server group ${desired.moniker} in ${desired.location.account}/${desired.location.region}"
+            else -> "Upserting server group ${desired.moniker} to $version in ${desired.location.account}/${desired.location.region}"
+          }
           log.info("Upserting server group using task: {}", job)
 
-          async {
+          val result = async {
             taskLauncher.submitJob(
               resource = resource,
-              description = "Upsert server group ${desired.moniker} in ${desired.location.account}/${desired.location.region}",
+              description = description,
               correlationId = "${resource.id}:${desired.location.region}",
               job = job
             )
           }
+
+          tags.forEach { tag ->
+            publisher.publishEvent(ArtifactVersionDeploying(
```

Comment:
Why one event for every tag? We just went through all this logic above to pick the one version string to use. Shouldn't we use that to generate a single event here?

Reply:
BTW, whatever the case, there should be tests covering the expected behavior when there are multiple tags for a digest.

Reply:
Well, we know what digest we're deploying. But, for docker images, there can be multiple tags for each digest. Say you publish your new image, and you tag it with the actual version `1.1.1`, and the `latest` tag happens automatically, but you're also doing some testing so you tag it `my-test-v1.1.1`. We have no way of knowing here which version you care about for managed delivery's sake. So the `getTags` list for this digest will return [`1.1.1`, `my-test-v1.1.1`, `latest`]. We filter out `latest`, cause we never care about that. But, in this context, how can we know which version string is the one that matches the format we're tracking, and which is the format that we can discard? `.first()` is meaningless. So, we send out an event for each version. In the handler for this event, we only mark as `DEPLOYING` if it's already been approved for the environment (which is how we ignore versions we're not explicitly caring about).

Reply:
Added tests for each publishing